src/php/datetime/strtotime.js

```diff
@@ -1,300 +1,1144 @@
+    this.m += this.rm
+    this.d += this.rd
+
+    this.h += this.rh
+    this.i += this.ri
+    this.s += this.rs
+    this.f += this.rf
+
+    this.ry = this.rm = this.rd = 0
+    this.rh = this.ri = this.rs = this.rf = 0
+
+    let result = new Date(relativeTo.getTime())
+    // since Date constructor treats years <= 99 as 1900+
+    // it can't be used, thus this weird way
+    result.setFullYear(this.y, this.m, this.d)
+    result.setHours(this.h, this.i, this.s, this.f)
+
+    // note: this is done twice in PHP
+    // early when processing special relatives
+    // and late
+    // todo: check if the logic can be reduced
+    // to just one time action
+    switch (this.firstOrLastDayOfMonth) {
+      case 1:
+        result.setDate(1)
+        break
+      case -1:
+        result.setMonth(result.getMonth() + 1, 0)
+        break
+    }
+
+    // adjust timezone
+    if (!isNaN(this.z) && result.getTimezoneOffset() !== this.z) {
+      result.setUTCFullYear(
+        result.getFullYear(),
+        result.getMonth(),
+        result.getDate())
+
+      result.setUTCHours(
+        result.getHours(),
+        result.getMinutes() + this.z,
+        result.getSeconds(),
+        result.getMilliseconds())
+    }
+
+    return result
+  }
+}
+
+module.exports = function strtotime (str, now) {
+  //       discuss at: http://locutus.io/php/strtotime/
+  //      original by: Caio Ariede (http://caioariede.com)
+  //      improved by: Kevin van Zonneveld (http://kvz.io)
+  //      improved by: Caio Ariede (http://caioariede.com)
+  //      improved by: A. Matías Quezada (http://amatiasq.com)
+  //      improved by: preuter
+  //      improved by: Brett Zamir (http://brett-zamir.me)
+  //      improved by: Mirko Faber
+  //         input by: David
+  //      bugfixed by: Wagner B. Soares
+  //      bugfixed by: Artur Tchernychev
+  //      bugfixed by: Stephan Bösch-Plepelits (http://github.com/plepe)
+  // reimplemented by: Rafał Kukawski
+  //           note 1: Examples all have a fixed timestamp to prevent
+  //           note 1: tests to fail because of variable time(zones)
+  //        example 1: strtotime('+1 day', 1129633200)
+  //        returns 1: 1129719600
+  //        example 2: strtotime('+1 week 2 days 4 hours 2 seconds', 1129633200)
+  //        returns 2: 1130425202
+  //        example 3: strtotime('last month', 1129633200)
+  //        returns 3: 1127041200
+  //        example 4: strtotime('2009-05-04 08:30:00+00')
+  //        returns 4: 1241425800
+  //        example 5: strtotime('2009-05-04 08:30:00+02:00')
+  //        returns 5: 1241418600
+  if (now == null) {
+    now = Math.floor(Date.now() / 1000)
   }
 
-  for (i = 0, len = match.length; i < len; i++) {
-    if (!process(match[i])) {
-      return fail
+  // the rule order is very fragile
+  // as many formats are similar to others
+  // so small change can cause
+  // input misinterpretation
+  const rules = [
+    formats.yesterday,
+    formats.now,
```

Comment:
Might you avoid the repetition of `formats` here with object destructuring?  (All of my comments are throw-away, so don't feel obligated to implement any of them...)

Reply:
Do you mean something like

    const { yesterday, now, ..., whitespace, any } = formats
    const rules = [ yesterday, now, ..., whitespace, any ]

?

Reply:
Yes, that, and if it makes sense, doing such destructuring within the function declaration, e.g. (one can use three periods (`...`) for gathering all other properties too):

```js
someMethod ({yesterday, now, whitespace, any, ...otherArgs}) {
    const rules = [ yesterday, now, ..., whitespace, any ]

}

```